dubbo-registry/dubbo-registry-consul/src/main/java/org/apache/dubbo/registry/consul/ConsulRegistry.java

```diff
@@ -0,0 +1,300 @@
+    private static final String WATCH_TIMEOUT = "consul-watch-timeout";
+    private static final String CHECK_INTERVAL = "consul-check-interval";
+    private static final String CHECK_TIMEOUT = "consul-check-timeout";
+    private static final String DEREGISTER_AFTER = "consul-deregister-critical-service-after";
+
+    private static final int DEFAULT_PORT = 8500;
+    // default watch timeout in millisecond
+    private static final int DEFAULT_WATCH_TIMEOUT = 60 * 1000;
+    // default tcp check interval
+    private static final String DEFAULT_CHECK_INTERVAL = "10s";
+    // default tcp check timeout
+    private static final String DEFAULT_CHECK_TIMEOUT = "1s";
+    // default deregister critical server after
+    private static final String DEFAULT_DEREGISTER_TIME = "20s";
+
+    private ConsulClient client;
+
+    private ExecutorService notifierExecutor = newCachedThreadPool(
+            new NamedThreadFactory("dubbo-consul-notifier", true));
+    private ConcurrentMap<URL, ConsulNotifier> notifiers = new ConcurrentHashMap<>();
+
+    public ConsulRegistry(URL url) {
+        super(url);
+        String host = url.getHost();
+        int port = url.getPort() != 0 ? url.getPort() : DEFAULT_PORT;
+        client = new ConsulClient(host, port);
+    }
+
+    @Override
+    public void register(URL url) {
+        if (isConsumerSide(url)) {
+            return;
+        }
+
+        super.register(url);
+    }
+
+    @Override
+    public void doRegister(URL url) {
+        client.agentServiceRegister(buildService(url));
+    }
+
+    @Override
+    public void unregister(URL url) {
+        if (isConsumerSide(url)) {
+            return;
+        }
+
+        super.unregister(url);
+    }
+
+    @Override
+    public void doUnregister(URL url) {
+        client.agentServiceDeregister(buildId(url));
+    }
+
+    @Override
+    public void subscribe(URL url, NotifyListener listener) {
+        if (isProviderSide(url)) {
+            return;
+        }
+
+        super.subscribe(url, listener);
+    }
+
+    @Override
+    public void doSubscribe(URL url, NotifyListener listener) {
+        Long index;
+        List<URL> urls;
+        if (ANY_VALUE.equals(url.getServiceInterface())) {
+            Response<Map<String, List<String>>> response = getAllServices(-1, buildWatchTimeout(url));
+            index = response.getConsulIndex();
+            List<HealthService> services = getHealthServices(response.getValue());
+            urls = convert(services);
+        } else {
+            String service = url.getServiceKey();
+            Response<List<HealthService>> response = getHealthServices(service, -1, buildWatchTimeout(url));
+            index = response.getConsulIndex();
+            urls = convert(response.getValue());
+        }
+
+        notify(url, listener, urls);
+        ConsulNotifier notifier = notifiers.computeIfAbsent(url, k -> new ConsulNotifier(url, index));
+        notifierExecutor.submit(notifier);
```

Comment:
I think the thread model here can be optimized, two potential problems:
1. busy looping will cost too much cpu.
2. One-thread-per-service will generate too many threads.

Reply:
![aaaaa](https://user-images.githubusercontent.com/18097545/54019981-9bd16700-41c7-11e9-9efa-3d694509598b.jpg)

Reply:
Above is the thread dump when subscribing to 2 services.

Reply:
Since consul's fetch API is pull-based  and with a blocking model, I think it's ok to keep the current thread model.